### ISSUE TITLE
AppRepo form currently requires fetching secrets from k8s API server

### ISSUE THREAD

**Issue by absoludity** [kubeapps/kubeapps#3922]
(I'm using this issue as a bit of a rubber-duck, but am keen for any other input too).

### Description:

I'm currently working on #3896 to remove the remaining dashboard calls to the K8s API server, in particular, where the Dashboard fetches secrets.

The only place where the dashboard fetches secrets is when editing an AppRepository via the AppRepo form, populating any credential field that had been set with the parsed secret value, such as:

![edit-repo-with-basic-auth](https://user-images.githubusercontent.com/497518/145339271-f61503cd-7c3c-41fc-be25-90d92ed8959e.png)

All of basic, bearer and custom options result in an opaque K8s secret with the data:

```
data:
  authorizationHeader: <b64-encoded-data-for-authorization-header>
```

rather than using secrets with a specific type ([such as `kubernetes.io/basic`](https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret)). So once the secret is created, there is nothing relating the type of authorization that it represents for the app repo other than parsing the actual secret data, which is [what the dashboard currently does :/ ](https://github.com/kubeapps/kubeapps/blob/v2.4.2/dashboard/src/components/Config/AppRepoList/AppRepoForm.tsx#L128-L151).

The constraints that I believe I have are:

- We don't want to add a GRPC endpoint to the resources plugin that just allows fetching secrets using the users credential (not far from proxying to the API server)
- We don't want to make too much change on the existing app repository API endpoint in the current backend, since replacing it with the apis server is in the works (#3496 )
- We do want to remove the dashboard's access to the k8s API server.

The main options I can see to move forward here are:

1. Update the dashboard to use the `resources.GetSecretNames` endpoint as I had planned, which includes the secret type for each name, then
2. Potentially change the UX experience so that you can't see previously entered secret values and are instead only able to replace the value with a new one when editing an AppRepository,

Or

3. Keep the same UX so that you can view existing credentials, by returning the secret data (both `authorizationHeader` and `ca.crt` when they exist) as part of the app repository from the backend for now.

AFAICT, if I implement 3 now, we will still need the `GetSecretNames` to populate the dropdown for the docker-config json's for image pull secrets. Anyway, I'll sleep on this and pick it up tomorrow.

**Comment by castelblanque** [comment]
For better security, secrets values shouldn't be accessible after creation. I would go for option 1+2, where user will not see existing secret value after creation and only replacing it with a new value is possible. Treatment is similar to passwords basically 😄 .
Now that `GetSecretNames` is in place, we are closer. Only downside is that we need UX changes.

**Comment by absoludity** [comment]
> For better security, secrets values shouldn't be accessible after creation. I would go for option 1+2, where user will not see existing secret value after creation and only replacing it with a new value is possible. Treatment is similar to passwords basically smile . Now that `GetSecretNames` is in place, we are closer. Only downside is that we need UX changes.

Thanks Rafa. Yeah, I agree that we should end up there, I'm just a bit wary of doing much to change this right now, given the planned changes here. I'm not going to get to do this before my PTO, so I'll see what the others think and tackle this when back.

**Comment by antgamdia** [comment]
Thanks for the changes for replacing the dashboard calls to the k8s server. Even if a proper refactor is the ideal option, as you mentioned, Rafa, I think our top priority now is getting rid of these direct dashboard->k8s API calls. In this context, although `3` is just shifting the problem from one component to another, it really solves our short-term issue.
Given that we plan to refactor the whole AppRepo UX thing, I believe we should go with the easiest solution for now.

**Comment by castelblanque** [comment]
Fine by me, we can make it in phases :)
For now we could implement solution `3` so that we have a quick win and schedule at some point the `1+2` option.

**Comment by absoludity** [comment]
Thanks for the input. I'll need to do this now to unblock #3896, so I'll go with (3) to unblock and we can update to (1) and (2) as part of the app repository refactoring.

**Comment by absoludity** [comment]
Implemented option 3 with #4027 and #4028.
